File: haskell/Puzzles/Day2.hs
module Puzzles.Day2 (puzzle) where

import Pre

import Data.Text qualified as T

-- this is somewhat slow? other than 4 (and to some extent 8), others are all very fast
-- generate repetitions up front instead? see aoc.md
-- do same for Rust
-- also look at Belle's micro-optimisations, though possibly irrelevant if I make that algorithm change
puzzle :: Puzzle
puzzle =
    Puzzle
        { number = 2
        , parser = const $ (<* newline) $ ((,) <$> (decimal <* char '-') <*> decimal) `sepBy` (char ',')
        , parts =
            ( sum
                . concatMap
                    (mapMaybe (\n -> guard (isRepetition2 n) $> n) . uncurry enumFromTo)
            )
                /\ ( sum
                        . concatMap
                            (mapMaybe (\n -> guard (isRepetitionN n) $> n) . uncurry enumFromTo)
                   )
                /\ nil
        , extraTests = mempty
        }

newtype ID = ID Int
    deriving newtype (Eq, Ord, Show, Num, Enum, NFData)

isRepetition2 :: ID -> Bool
isRepetition2 (T.show -> n) = case T.length n `divMod` 2 of
    (d, 0) -> equalChunks n d
    _ -> False

isRepetitionN :: ID -> Bool
isRepetitionN (T.show -> n) = flip any [1 .. T.length n `div` 2] $ equalChunks n

equalChunks :: Text -> Int -> Bool
equalChunks n i = case T.chunksOf i n of
    [] -> True
    x : xs -> all (== x) xs

File: rust/puzzles/day2.rs
use crate::puzzle::Puzzle;
use nom::{
    Parser,
    character::complete::{char, newline, usize},
    combinator::eof,
    error::Error,
    multi::separated_list1,
    sequence::{separated_pair, terminated},
};

pub const PUZZLE: Puzzle<Vec<(usize, usize)>, 2> = Puzzle {
    number: 2,
    parser: |_, input| {
        terminated::<_, _, Error<&str>, _, _>(
            terminated(
                separated_list1(char(','), separated_pair(usize, char('-'), usize)),
                newline,
            ),
            eof,
        )
        .parse(input)
        .unwrap()
        .1
    },
    parts: [
        |input| {
            input
                .into_iter()
                .flat_map(|(l, u)| {
                    (*l..(u + 1)).flat_map(|n| if is_repetition_2(n) { Some(n) } else { None })
                })
                .sum::<usize>()
                .to_string()
        },
        |input| {
            input
                .into_iter()
                .flat_map(|(l, u)| {
                    (*l..(u + 1)).flat_map(|n| if is_repetition_n(n) { Some(n) } else { None })
                })
                .sum::<usize>()
                .to_string()
        },
    ],
};

fn is_repetition_2(n: usize) -> bool {
    let n = n.to_string();
    let l = n.len();
    let d = l / 2;
    let r = l % 2;
    if r == 0 { equal_chunks(&n, d) } else { false }
}

fn is_repetition_n(n: usize) -> bool {
    let n = n.to_string();
    let l = n.len();
    let d = l / 2;
    (1..(d + 1)).any(|i| equal_chunks(&n, i))
}

fn equal_chunks(n: &String, i: usize) -> bool {
    let chars = n.chars().collect::<Vec<char>>();
    let mut chunks = chars.chunks(i);
    match chunks.next() {
        None => true,
        Some(x) => chunks.all(|y| y == x),
    }
}
